websockets/cext.c

```diff
@@ -0,0 +1,128 @@
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+
+#include <stdlib.h> /* getenv */
+
+#ifndef HAS_SSE2
+/* setup HAS_SSE2 macro according to compiler predefined macros */
+#ifdef __SSE2__
+#define HAS_SSE2 1
+#else
+#define HAS_SSE2 0
+#endif
+#endif /* not defined HAS_SSE2 */
+
+#if HAS_SSE2
+#include <emmintrin.h>
+#endif
+
+#ifndef HAS_FAST_UNALIGNED_ACCESS
+/* This is not enabled by default */
+/* build with CFLAGS="-DHAS_FAST_UNALIGNED_ACCESS=1" to enable */
```

Comment:
Let's see,

`HAS_FAST_UNALIGNED_ACCESS` was there because there's no normative information about alignment of buffers returned by `PyArg_ParseTuple(args, "y#y#", &in, &len, &mask, &mask_len)`and `out = PyBytes_AS_STRING(result)`.
Enabling it allowed to unconditionally cast any of those pointers to `uint32_t*` or `uint64_t*` and using those pointers to get/store values directly.
If one of the pointers is not aligned on an 8byte boundary (e.g. `out = 0x7`), then 3 things could happen:
1. the platform supports unaligned access and it's fast => no worries there
2. the platform doesn't support unaligned access then `*(uint64_t*)(out + 0) = value` would crash because we try to store an `uint64_t` value at an address that's not valid for an `uint64_t` which require 8byte alignment of the address.
3. the platform supports unaligned access but those are slow. Well, then it's just slow. All x86/x64 now have fast unaligned access. ARM doesn't support unaligned access natively and the linux kernel traps them to return or store the correct value by doing multiple access => this is dead slow.

What is done now:
I assume that pointers will always be aligned on an 8byte boundary and so we provide the fastest implementation only for that case.
Reading the code of cpython, the size of the header for `PyBytesObject` will be 32bytes on 64bit platforms (that's your 33bytes which is 32+1 element in the buffer which is always NUL terminated) and 16 bytes on 32bit platforms. Now, given the fact that a `PyBytesObject`is itself allocated with an allocator respecting an 8byte boundary (either system malloc, cpython custom allocator) and the size of the header is a multiple of 8, the assumption that those buffers are always 8byte aligned is true in that case.
For cases that would not be covered by this assumption, the access to unaligned pointers is simply disabled falling back to unoptimized case but preventing any crash that could happen.
This test is `if ((((uint32_t)in | (uint32_t)out | (uint32_t)mask) & 7U) == 0) {` 

I hope this is somewhat clearer.